[TG-5572] Add option to restrict numerical input values

majakusber opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

Adds command line option `--java-assume-input-limits [int:int]`. Does not work for arrays yet.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier commented on jbmc/regression/jbmc/assume-input-limits/My.java:
```diff
@@ -0,0 +1,131 @@
+
+  int[] intArrField;
+
+  public static void numericalArg(byte byteArg, short shortArg, int intArg, long longArg, float floatArg, double doubleArg) {
```
this line is too long

majakusber replied:
Shortened

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1011,6 +1011,28 @@ void java_object_factoryt::gen_nondet_init(
 
     assignments.add(assign);
+    // add assumes to obey numerical limits (given via java-assume-input-limits)
+    if(type.id() != ID_c_bool && type.id() != ID_char)
```
:pick: I would prefer `type != java_bool_type() && type != java_char_type()`

thk123 replied:
This behaviour should be documented in the command line options help

majakusber replied:
Done and done

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1011,6 +1011,28 @@ void java_object_factoryt::gen_nondet_init(
+    // add assumes to obey numerical limits (given via java-assume-input-limits)
+    if(type.id() != ID_c_bool && type.id() != ID_char)
+    {
+      if(object_factory_parameters.assume_input_limits.lower.has_value())
```
:pick: maybe do `if(const auto &lower = ...)` so that you can use `lower.value()` below instead of the long expression

majakusber replied:
Done

romainbrenguier commented on src/util/object_factory_parameters.cpp:
```diff
@@ -46,6 +47,24 @@ void object_factory_parameterst::set(const optionst &options)
+    DATA_INVARIANT(
+      colon != std::string::npos && limits.front() == '[' &&
+        limits.back() == ']',
+      "java-assume-input-limits value should be pf the form [int:int]");
```
typo "pf" -> of

majakusber replied:
Corrected

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 01f3e63a3673746ccb08b09c0cec50d3ecba47db).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115054010

romainbrenguier commented on src/util/object_factory_parameters.cpp:
```diff
@@ -46,6 +47,24 @@ void object_factory_parameterst::set(const optionst &options)
+        temp.lower = unsafe_string2int(limits.substr(1, colon - 1));
+      if(colon < limits.size() - 2)
+        temp.upper = unsafe_string2int(
+          limits.substr(colon + 1, limits.size() - colon - 2));
```
I think using a regex `"\\[(\\d*):(\\d*)\\]"` would be a bit clearer and avoid having to think about indexes

majakusber replied:
Done

romainbrenguier commented on src/util/object_factory_parameters.h:
```diff
@@ -70,6 +71,14 @@ struct object_factory_parameterst
+    optionalt<int> lower;
+    optionalt<int> upper;
+  };
+  numeric_input_limitst assume_input_limits;
```
I'm not sure these two need to be grouped in a struct type, since the type itself doesn't seem to be used outside of this field.

thk123 replied:
I disagree - they are logically one thing (a range). However, does it make sense to have a lower but not an upper or vice versa? I'd probably make the whole thing optional if you only support having both or neither set

majakusber replied:
I also find it more understandable if they are grouped. 

Originally, I had the whole thing as optional but I think supporting only one limit could be quite useful (allows to for example force using only positive values). Alternatively, at one point I had default values for the limits - min and max long. However, asking whether the optional has a value is I think nicer than comparing the value to max long. I also thought of making the whole thing as well as the individual limits optional but that seems like an overkill :slightly_smiling_face:

romainbrenguier replied:
In that case "range" or "interval" should probably appear in the name. 
Rather that defining a new type you should reuse existing ones, here `integer_intervalt` would be the right one. It will avoid having to define a new type and you can make your code simpler by using `.empty()`.

majakusber replied:
Oh yes! That looks like exactly what I need, thank you :+1:

thk123 commented on src/util/object_factory_parameters.cpp:
```diff
@@ -46,6 +47,24 @@ void object_factory_parameterst::set(const optionst &options)
+  {
+    const auto &limits = options.get_option("java-assume-input-limits");
+    const auto &colon = limits.find(':');
+    DATA_INVARIANT(
```
:no_entry_sign: User errors should throw an exception rather than an invariant. Specifically `user_input_error_exceptiont`

majakusber replied:
Done, I used `invalid_command_line_argument_exceptiont`

thk123 commented on jbmc/regression/jbmc/assume-input-limits/My.java:
```diff
@@ -0,0 +1,131 @@
+  int[] intArrField;
+
+  public static void numericalArg(byte byteArg, short shortArg, int intArg, long longArg, float floatArg, double doubleArg) {
+    if (intArg > 3) {
```
:pick: put the ifs in the same order as the parameters

majakusber replied:
Done

thk123 commented on jbmc/regression/jbmc/assume-input-limits/My.java:
```diff
@@ -0,0 +1,131 @@
+    }
+  }
+
+  public void classArrArg(Other other, int [] intArr) {
```
:pick: Ideally this would go in the other (Array) commit to ease reviewing

majakusber replied:
Done

thk123 commented on jbmc/regression/jbmc/assume-input-limits/arg_assume_no_zero.desc:
```diff
@@ -0,0 +1,23 @@
+CORE
+My.class
+--function My.numericalArg --java-assume-input-limits [1:3]
```
:question: Not clear on what this test checks? There is no assert that can be reached by only using 0?

majakusber replied:
I changed the tests a bit - all desc files check for positive limits only `[1:3]` and there is a specific test fro checking that negative limit works too `arg_assume_neg`

thk123 reviewed (commented):
:no_entry_sign: Please add some tests for invalid format of parameter and check the exit code is `1` :white_check_mark:

tautschnig commented on src/util/object_factory_parameters.h:
```diff
@@ -70,6 +71,14 @@ struct object_factory_parameterst
+  struct numeric_input_limitst
+  {
+    optionalt<int> lower;
+    optionalt<int> upper;
```
Is `int` really the right type? I'd use `mp_integer` here as the type of the system JBMC is being compiled on is unrelated to the type that you want to talk about.

majakusber replied:
Good point, done!

majakusber commented:
@romainbrenguier @thk123 @tautschnig All comments addressed, please have another look. I left most of the changes in the `Fixups`/`Fixup` commits but will squash those before merging.

General question regarding the option name: would you prefer `--java-assume-input-limits` or `--java-assume-input-bounds`?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 56a84cfd4a5cd859723af9167bf1e36eb0ddc40c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115267886

romainbrenguier commented on src/util/object_factory_parameters.cpp:
```diff
@@ -46,6 +50,37 @@ void object_factory_parameterst::set(const optionst &options)
+  if(options.is_set("java-assume-input-limits"))
+  {
+    const auto &limits = options.get_option("java-assume-input-limits");
+    const std::regex limits_regex("\\[(-\\d+|\\d*):(-\\d+|\\d*)\\]");
```
:+1:

romainbrenguier commented on src/util/object_factory_parameters.h:
```diff
@@ -70,6 +72,14 @@ struct object_factory_parameterst
   /// Force string content to be ASCII printable characters when set to true.
   bool string_printable = false;
 
+  /// Force numerical primitive inputs to fall within the bounds
```
The documentation should also tell what the type represents not how it will be used.

Commit:
Add regression test for static fields

Using --nondet-static

Commit:
Rename option to java-assume-inputs-limit and use existing intervalt

The name now matches the related java-assume-inputs-non-null option and
better reflects the intention.

Instead of introducing new class to store the bounds, existing interval
implementation is used.

Commit:
Add regression test for java-assume-input-limits for arrays

Marked as KNOWNBUG, the option does not work for arrays yet

Commit:
Add command line option to define limits for numerical inputs

The option is of the form --java-assume-input-limits [int:int]

Commit:
Add regression tests for java-assume-input-limits

Commit:
Move integer_intervalt into a file in util

majakusber commented:
@tautschnig @thk123 Ready for re-review, as @romainbrenguier suggested I used `integer_intervalt` instead of my own struct, I renamed the option to match this and also to match the related `java-assume-inputs-non-null` option, and I added tests for static fields.

romainbrenguier commented on src/util/object_factory_parameters.h:
```diff
@@ -14,6 +14,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <util/irep.h>
+#include <util/optional.h>
+#include <analyses/interval_domain.h>
```
I think we shouldn't have dependencies of util on analyses, so the best solution is probably to move `interval_template.h` to  `util` (and we can drop "template" from the name)

majakusber replied:
Yep, moved... with a bunch of other files it needed...

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -56,6 +57,11 @@ Author: Daniel Kroening, kroening@kroening.com
   "                              entry point with null\n" /* NOLINT(*) */ \
+  " --java-assume-inputs-interval [L:U] or [L:] or [:U]\n" \
+  "                              never initialize numerical primitive-typed parameter\n" /* NOLINT(*) */ \
+  "                              (byte, short, int, long, float, double) to an entry\n" /* NOLINT(*) */ \
```
:bulb: I'd phrase this as the positive "only initialize primitive typed parameters to values within the specified range"

majakusber replied:
Rephrased

thk123 reviewed (approved):
lgtm

tautschnig reviewed (changes requested):
I'm afraid the move of files is a bit too far-reaching. It seems that all that you need is `integer_intervalt`, so I guess that should just be placed in a file of its own. The `ai*.h` files really shouldn't be moved.

romainbrenguier commented on unit/analyses/ai/ai.cpp:
```diff
@@ -12,7 +12,7 @@ Author: Diffblue Ltd.
 #include <testing-utils/use_catch.h>
 
-#include <analyses/ai.h>
+#include <util/ai.h>
```
the header file and cpp file shouldn't be in separate modules and I think ai should stay in analyses

majakusber commented:
I minimized the move :crossed_fingers:

romainbrenguier commented on src/util/integer_interval.h:
```diff
@@ -0,0 +1,18 @@
+#include "interval_template.h"
+#include "mp_arith.h"
+
+typedef interval_templatet<mp_integer> integer_intervalt;
```
:pick: `using` is preferred

majakusber replied:
Done

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f303e99ca7afca324b27498bc6548902db99d4fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115425794

majakusber commented:
@tautschnig Would you have time to give a full review? I need a code-owner :)

tautschnig commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1011,6 +1011,27 @@ void java_object_factoryt::gen_nondet_init(
     assignments.add(assign);
+    // add assumes to obey numerical limits (given via
+    // java-assume-inputs-interval)
+    if(type != java_boolean_type() && type != java_char_type())
```
Why is `java_char_type()` also special cased? It would seem that intervals still make sense for that type?

romainbrenguier replied:
it makes sense but we would probably want a different one, for instance if we use [0:10] for integers it would only give non-printable characters

majakusber replied:
I think that wouldn't be very intuitive for the user (definitely not for me :slightly_smiling_face: ). We have `--string-input-value` option which accepts a list of strings to use during test generation (it's not exactly the same though as what you meant).

thk123 replied:
A little forward visibility - a forthcoming feature will bound `java_char_type` when `--string-printable` is used (using the same range of characters).

tautschnig commented on src/util/object_factory_parameters.cpp:
```diff
@@ -46,6 +50,34 @@ void object_factory_parameterst::set(const optionst &options)
+    }
+    assume_inputs_interval = [&]() -> integer_intervalt {
+      integer_intervalt temp;
+      if(base_match[1] != "")
```
Why not `!base_match[1].empty()`

majakusber replied:
The type doesn't have `empty()` function. Note that because the regex allows empty string here, this would always be a match. I just need to check that what matched was an empty string (no bound).

tautschnig commented on src/util/object_factory_parameters.cpp:
```diff
@@ -46,6 +50,34 @@ void object_factory_parameterst::set(const optionst &options)
+      integer_intervalt temp;
+      if(base_match[1] != "")
+        temp.make_ge_than(string2integer(base_match[1]));
+      if(base_match[2] != "")
```
Why not `!base_match[2].empty()`
